### ISSUE TITLE
Auto-mount like AVFS (or better) for transparent navigation

### ISSUE THREAD

**Issue by amerlyq** [mxmlnkn/ratarmount#56]
Basically, I expected something AVFS-esque, but couldn't wrap my head how to achieve this with ratarmount.
What AVFS currently does:
* mount whole "/" to `~/.avfs` as FUSE
* when accessing archives under `~/.avfs/` (with `#` appended to path) it auto-mounts them to these paths e.g. `~/.avfs/tmp/bash-5.0.tar.gz#/`

Improvement for above workflow would be to auto-unmount these bindmounts too, when e.g. lsof shows no program is using files under mountpoint (however better method would be desirable).

Primary usecase: integration with filemanagers, to jump into VFS folder when trying to open/preview archive.
E.g. https://github.com/ranger/ranger/issues/456#issuecomment-797925011

**Comment by mxmlnkn** [comment]
Your requirements remind me of the setup mentioned in #25. The avfs feature seems kinda orthogonal to what ratarmount currently does. It feels more like ratarmount should be added as a plugin or addon or whatever to avfs.

I guess it wouldn't be impossible though to add some "mount on usage" feature to ratarmount. Especially, as it already supports bind-mounting and recursive mounting. With `ratarmount -r / ~/.avfs` you already should be able to mount all tar files you have. However, that would read all tar-files during mounting instead of reading them when actually needed. And adding new tar files at a later time wouldn't work. You would have to unmount and mount ratarmount again. On the second mount, ratarmount would only have to read the newly created tar files but it still is a hassle to remount on changes.

I think, I'd be ok with adding a `--lazy` recursive mount option. It wouldn't be so hard, I think. However, automatic unmounting would be a bit more sophisticated because I don't know how to keep track of opened file handles, I guess you could parse lsof but that doesn't feel right. Maybe, the FUSE open, opendir, release, and releasedir could be overwritten in order to track opened paths but that would need testing.

In the short-term, I don't see me having time for this. Merge requests are welcome though.

**Comment by amerlyq** [comment]
Some points:
* `--lazy` is definitely the way to go -- and for viable proof-of-concept it would be enough, if implemented.
* proper on-disk DB is not needed -- I don't mind waiting a little on first access to archive of reasonable size (<1GB), until its index generated and loaded into `:memory:` db or simply native structures. Of course option for persistence into some cache db with entries having TTL to keep it under reasonable size is welcomed.
* making ratarmount into addon/plugin of AVFS is less likely, considering frequency of AVFS sourceforge releases and its primary usecase being a library instead of FUSE -- moreover most of functionality related to archive viewing is overlapping.
* actually necessity for trailing `#/` to open archive in AVFS is the result of afterthought added FUSE support, it would be **much** better to directly represent archives as folders in VFS, because all we need -- is to jump into prepared path from filemanager, not access original file by similar path under VFS.
* automatic unmounting is only nice-to-have and not critical feature, I don't mind doing `umount -R` once in a while.
* on the other hand, when using FUSE -- there is no need to even bindmount anything in nested folders -- we could directly provide filesystem representation inside archive. That leaves root mount point the only mounted folder -- and there is no need to unmount it at all. Isn't `--recursive` already works as such, or am I mistaken?

So, yep, `--lazy` with `:memory:` would be good even by themselves, and it doesn't seem like too much of architectural change.
However it still may require significant and contradictory code change to accomodate both current and this new workflows.

> Merge requests are welcome though.

It would be the best if you could kindly spend under half a hour to list places in source code most likely to be touched by aforementioned changes.
Initial navigation in the unknown codebase is a pain, because it's hard to understand which parts are truly unrelated and so don't deserve to spend any attention, time and efforts, and which places are very important but experience lack of earnestness :)

**Comment by mxmlnkn** [comment]
> automatic unmounting is only nice-to-have and not critical feature, I don't mind doing `umount -R` once in a while.

That also wouldn't work with the current ratarmount setup and the extension I have in mind. Everything would be under one mount point and ratarmount would manage "sub mount points" without the system knowing anything about it.

> on the other hand, when using FUSE -- there is no need to even bindmount anything in nested folders -- we could directly provide filesystem representation inside archive. That leaves root mount point the only mounted folder -- and there is no need to unmount it at all. Isn't `--recursive` already works as such, or am I mistaken?

You are completely right and that's what I wanted to say above.

> It would be the best if you could kindly spend under half a hour to list places in source code most likely to be touched by aforementioned changes.

Let's see. The ratarmount structure is relatively short and everything is in ratarmount.py in order to support simply copying this script in contrast to a full install. Structure:

 - Main with argument parsing
   - TarMount implements the fusepy interface and manages multiple mount sources (folders or multiple TARs)
     - FolderMountSource implements the bind mounting of folders and the recursive mounting of TARs in those folders
     - SQLiteIndexedTar implements the random access to TARs

The lazy mounting would have to be implemented into FolderMountSource. Currently, it creates `self.mountedTars` and `self.rootFileInfos` in the [constructor](https://github.com/mxmlnkn/ratarmount/blob/master/ratarmount.py#L1603-L1629). This would have to be done dynamically during the accesses in `getFileInfo` and `_findMountedTar`.

Actually, I found some time to implement this, see e875dddb0ccd3925ef4e451862f26daa46da33b1. You'll have to see whether it suffices. Try it with:

```bash
pip install git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount
```

I'm sure it's not perfect and the CI error looks weird, too. But it should work more or less.

**Comment by amerlyq** [comment]
Cool, I have tested it. 

* somehow it mounts all archives in folder on entering, and creates .sqlite dbs for all of them, despite I'm interested only in some.
* navigating archive with ~500 files is unberably slow, no problem with <100 files, or when using AVFS.

But it somehow works, and it's good already :)

**Comment by mxmlnkn** [comment]
> somehow it mounts all archives in folder on entering, and creates .sqlite dbs for all of them, despite I'm interested only in some.

Well, that's expected. I went with a quickfix, which implements it on the first stat, which is done during listing of the folder in which the TAR resides in. With just a tad of work, it could be changed to be mounted during listDir of that TAR archive.

> navigating archive with ~500 files is unberably slow, no problem with <100 files, or when using AVFS.

That sounds bad. In the first place, I began with ratarmount in order to be faster than alternatives. However, the current bind-mounding implementation has some known performance issues in the code like checking against all mounted TAR mount points on each stat and listdir.

Could you maybe test whether that performance issue remains when only mounting the TAR itself without any of the extended features like recursive mounting or bind mounting folders and so on. If the performance issue still remains, could you give me a minimal example? I mean, generally, I did test it with TARs with millions of files at some point, so it should work to some extent. I know that running `find` for a mounted TAR is [slower when compared to archivemount](https://raw.githubusercontent.com/mxmlnkn/ratarmount/master/benchmarks/plots/archivemount-comparison.png), But I wouldn't call it "unbearable" slow. At least the complexity scaling is the same.

**Comment by amerlyq** [comment]
Some numbers and narrowing the scope:
I compressed 4000 files (12-20bytes each) in my `~/.config/mpv/watch_later` and scrolled 300 first ones in `ranger` (preview disabled) and `nnn` with stopwatch :).
```
[ranger]
original=7.73s
AVFS=7.91s
ratarmount=12.21s

[nnn]
original=7.86s
AVFS=7.79s
ratarmount=7.99s
```
So, the difference here is the number of `stat()` calls. Yep, ratarmount is slow on that aspect, you had warned.

> navigating archive with ~500 files is unberably slow, no problem with <100 files, or when using AVFS.

I triaged and this problem had occured when all these conditions are met:
0) using jq.tar.zst with single frame.
1) using "flattening" feature from `ranger`, i.e. representing all files in recursive directories as single list, which, I assume, was implemented half-heartedly and requires tons of jumping back and forth with collateral `stat` calls; 
2) enabling lazy file preview, which produces 10+ background threads to process each file under cursor into some kind of info summary.
Under all these conditions scrolling 300 files:
```
unpacked=12.47s
AVFS=16.24s
ratarmount=22.26s
```
Most impactful obviously lazy preview thingy. 
But to be fair, I don't have any expectation from python, so despite having room for improvents... it's actually ok?

**Comment by mxmlnkn** [comment]
Thank your for your extensive tests. Unfortunately, I'm of the same opinion as you state in your closing remark. A slowdown of factor 2 doesn't seem all that bad to me considering it's written in Python. The most important for me during creation of ratarmount was the speedup when jumping to TAR blocks when reading files and the loadable index.

I already tried to speed stat calls by caching results from the SQL queries into Python dictionaries but I couldn't measure any performance improvements. My takeaway back then was that this Python implementation simply has too many layers - like fuse, fusepy, and the python interpreter itself - stacked ontop of each other and the slowdown is had to improve upon.

It might help and actually might be possible to reimplement ratarmount in native C++ but that would be quite some work and would require managing those dependencies like zstd, bz2, gzip, sql, json and argument parsing. With pip and pypi and the python standad library, it was quite easy to use all those things.

**Comment by amerlyq** [comment]
> reimplement ratarmount in native C++

meh, I like ratarmount exactly for python, so I can deeply integrate it with some inner machineria of e.g. ranger and reduce number of layers.

> managing those dependencies like zstd, bz2, gzip, sql, json and argument parsing

Conan is good for building that, we still have Snap for everybody, and AUR in archlinux for me, so no problem, but python is simply more hackable for current state of project.

> A slowdown of factor 2 doesn't seem all that bad to me considering it's written in Python

Looking as AVFS creates new threads for each file access to accomodate decompressing in parallel, maybe we still have the same low hanging fruit for ratarmount too? I remember you must explicitly enable multithreading in FUSE and write code somewhat differently for that to work, though.

Ok, for now current state of things on lazy-opening side is good enough for me to use.
"Mounting" everything when entering the folder is viable.
Every other discrepancy is worth its own separate issue.

The only action item from my standpoint -- is to prevent creation of sqlite databases on disk, as instead of several large archives I have ~500+ archives of size 1MB..80MB, and whole usefulness of persistent sqlite between reboots is lost on them. 
Direct passing of single memory-only db for everything didn't worked.
```
ratarmount --index-file ':memory:' --lazy -r / /run/user/1000/ratarmount/
```

Do you have in mind any other action items, which must be achieved before this issue can be closed?

**Comment by mxmlnkn** [comment]
> Looking as AVFS creates new threads for each file access to accomodate decompressing in parallel, maybe we still have the same low hanging fruit for ratarmount too? I remember you must explicitly enable multithreading in FUSE and write code somewhat differently for that to work, though.

You might be right. I don't know how to activate that, though. And with Python, you also have the global interpreter lock, which might hinter multi-threading.

> The only action item from my standpoint -- is to prevent creation of sqlite databases on disk, as instead of several large archives I have ~500+ archives of size 1MB..80MB, and whole usefulness of persistent sqlite between reboots is lost on them.
> Direct passing of single memory-only db for everything didn't worked.

I think it will be ignored for multiple sources because every source has its own database and it becomes ambiguous which of the sources should use the given name. But `:memory:` seems like a index file name, which could and should be used for all sources if there are multiple given.

> Do you have in mind any other action items, which must be achieved before this issue can be closed?

 - Normally, I close issues when merging the fix into master. For that, I need to find out why my CI fails, maybe it's even one of the other commits, which causes it.
 - Maybe also implement that `:memory:` index file name works.

Aside from that I don't see any blockers right now.

**Comment by amerlyq** [comment]
> global interpreter lock, which might hinter multi-threading

Forking is also variant -- after all readonly decryption can occur in parallel and isn't expected to share any state.

> :memory: seems like a index file name

Let's align our understanding: `:memory:` is a special database path, which when passed to `sqlite3` creates memory-only db (like regular one but without disk backing), identified only by current sqlite db instance, and which will be auto-destroyed when all connection (fd) are closed. So you can have parallel connections even in this case to accomodate above forking case.

Therefore if you open `:memory:` db for each archive -- it will be different db altogether for each connection, no need to rework architecture and merge all indexes into one db (however it may be interesting future task to make this virtual index optionally persistent).
And if you add flag to pass this magic string to your db code -- it may "just work (c)" in several lines of code, and won't create any on-disk db.

Maybe the only performance downside will occur if you often close/open db connections itself -- which will result in necessity to regenerate index for new `:memory:` instance each time -- still if connection is held until I navigate archive under FUSE -- it's enough.

**Comment by mxmlnkn** [comment]
> Let's align our understanding

I see nothing wrong with what you said, including the performance downside.

**Comment by mxmlnkn** [comment]
I think I'll close this for now with ed68786. Feel free to reopen or open another issue if the missing auto-"unmount" leads to some kind of performance problems. As discussed, the recursive mounting is done internally so it should be a bit more lightweight than a full system mount. One problem I can think of might be that the upper limit of open file handles might get hit when recursively mounting a thousand or so archives. That limit can be increased with `ulimit` though.
